src/main/java/io/cryostat/net/web/http/api/beta/graph/GraphModule.java

```diff
@@ -0,0 +1,721 @@
+                                                if (settings.containsKey("maxAge")) {
+                                                    builder =
+                                                            builder.maxAge(
+                                                                    (Long) settings.get("maxAge"));
+                                                }
+                                                if (settings.containsKey("maxSize")) {
+                                                    builder =
+                                                            builder.maxSize(
+                                                                    (Long) settings.get("maxSize"));
+                                                }
+                                                IRecordingDescriptor desc =
+                                                        helper.startRecording(
+                                                                cd,
+                                                                builder.build(),
+                                                                (String) settings.get("template"),
+                                                                TemplateType.valueOf(
+                                                                        ((String)
+                                                                                        settings
+                                                                                                .get(
+                                                                                                        "templateType"))
+                                                                                .toUpperCase()));
+                                                return null;
+                                            },
+                                            true);
+                                } catch (Exception e) {
+                                    logger.error(e);
+                                    exceptions.add(e);
+                                }
+                            });
+            if (!exceptions.isEmpty()) {
+                throw new BatchedExceptions(exceptions);
+            }
+
+            return matches;
+        };
+    }
+
+    @Provides
+    @Singleton
+    @Named("startRecordingByLabels")
+    static DataFetcher<List<TargetNode>> provideStartRecordingByLabelsFetcher(
+            PlatformClient client,
+            TargetConnectionManager tcm,
+            RecordingTargetHelper helper,
+            RecordingOptionsBuilderFactory recordingOptionsBuilderFactory,
+            CredentialsManager credentialsManager,
+            Logger logger) {
+        return env -> {
+            Map<String, Object> settings = env.getArgument("recording");
+            List<Map<String, String>> labelSelectors = env.getArgument("labels");
+
+            EnvironmentNode root = client.getDiscoveryTree();
```

Comment:
> For performance reasons we might want to keep an in-memory model of the whole Kubernetes deployment cached for some time and periodically refresh it

Revisiting this, is the list of active and archived recordings for each target considered part of this deployment model? If so, would we need to refresh this cache every time a mutation operates on your targets? For example, if you accidentally started several recordings and wanted to stop them all, the `provideStopRecordingByLabelsFetcher` would need to know what recordings were started on those targets.

Reply:
The active and archived recordings aren't things that Kubernetes knows about, so mutations to those resources doesn't require a re-query from Kubernetes for the deployment model. We would be able to just update our in-memory model of the active and archived recordings that we know about. We may still want to periodically refresh those models to ensure that our model isn't out of sync with reality, particularly for active recordings since it's possible that some other JMX client has mutated the recordings on some targets, or it's possible that a target has been restarted and lost its active recordings (if they are not started by JVM flags on the target, instead started by Cryostat either manually or via automated rules).

Anyway, your example of wanting to stop specific recordings within a set of targets is an interesting one, and I think very worthy of consideration. This also links back to the other comment chain and #709. Maybe we want to add metadata to allow users to attach annotations and labels to the recordings themselves. Then a user can make a single query like "start a recording with annotation `com.app/jfrProfiling` on all targets with label `app=checkoutService`", and later they can say "stop recordings with annotation `com.app/jfrProfiling` on all targets with label `app=checkoutService`".

As it is right now, there is an implicit underlying assumption that users only really use one recording at a time per target. It is supported to have multiple because JFR supports this very cleanly, but it's true that Cryostat's API tends to operate on things at the target level and all recordings belonging to a target.

Reply:
> We would be able to just update our in-memory model of the active and archived recordings that we know about.

This makes sense, thanks.

> As it is right now, there is an implicit underlying assumption that users only really use one recording at a time per target.

Do you know of any use cases that would involve running more than one recording at once? For example, starting a Profiling recording for a short duration while an older Continuous recording runs in the background?
It seems like adding metadata would be worthwhile if you needed to differentiate between multiple running recordings. If there's only one running recording, you could just ask Cryostat to stop all recordings on the JVM.

Reply:
Yes, your example is a very valid one and one that we should probably expect to happen quite a lot. A Continuous template recording should be running at all times on most services, and a Profiling template recording would be used to help diagnose and troubleshoot when a specific problem is detected using the information provided by the Continuous recording. We shouldn't ask the user to stop the Continuous recording before starting the Profiling recording, and we don't want to do any of that manual juggling ourselves (stop the Continuous, start the Profiling, start the Continuous again when that's done) because it's unnecessary overhead and bookkeeping.

In other words, the implicit underlying assumption is not a good one. But it's the nature of how the API has been built so far, where recordings are only really differentiated by the target they belong to and their human-friendly name. I don't think we want to start enforcing specific names for recordings, so this metadata idea is pretty powerful and important.